### PR TITLE
Add sh.ppy.osulazer

### PR DESCRIPTION
**Pull request by BiteDasher**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions.
`--device=all` is needed for the functioning of things like a graphics tablet/Wii Remote
`--filesystem=xdg-{music,download,video,pictures}` is needed for the import of files in beatmap editor
`--filesystem=~/.wine` is needed for the migration of beatmaps from the Windows version of osu! (if installed)
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [ ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge) (owned github/BiteDasher/flatpak-osu-lazer, not github/ppy/osu).
- [ ] Any additional patches or files have been submitted to the upstream projects concerned. *(updater.patch is needed to disable the behavior of notifications about a new version of the game and an offer to download an update in the form of an AppImage. Also removes the "Check for updates" button from the settings)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by BiteDasher** [comment]
bot, build sh.ppy.osulazer

**Comment by BiteDasher** [comment]
bot, build sh.ppy.osulazer

**Comment by BiteDasher** [comment]
image
Icons of MIME types in nautilus seems blurry. Also, this happens if use symlinks. Gonna undo this.

**Comment by BiteDasher** [comment]
bot, build sh.ppy.osulazer

**Comment by BiteDasher** [comment]
One more question, the game resources are licensed under *CC-BY-NC 4.0*.  Can i add another `project_license` tag to appdata? Is this will be valid?

**Comment by BiteDasher** [comment]
bot, build sh.ppy.osulazer

**Comment by BiteDasher** [comment]
bot, build sh.ppy.osulazer

**Comment by BiteDasher** [comment]
image
If I remove sh.ppy.osulazer.png from /app/share/icons/Adwaita/*/apps, this happens. Gonna fix it.

**Comment by BiteDasher** [comment]
Due to some mechanism in GNOME, application icon is being appear as mimetype icon. Because of this, the game icon in the menu is displayed incorrectly.

**Comment by BiteDasher** [comment]
bot. build sh.ppy.osulazer

**Comment by BiteDasher** [comment]
Everything works OK if I install `mime` logo as `/app/share/icons/*/*/mimetypes/x-osu-beatmap`. I will be glad if you can give me advice on how to put something other than `$FLATPAK_ID` in `/app/share/icons`

**Comment by BiteDasher** [comment]
bot, build sh.ppy.osulazer

**Comment by BiteDasher** [comment]
> Everything works OK if I install `mime` logo as `/app/share/icons/*/*/mimetypes/x-osu-beatmap`. I will be glad if you can give me advice on how to put something other than `$FLATPAK_ID` in `/app/share/icons`, but this is not necessary.

@hfiguiere any ideas? I can't found anything about this

**Comment by BiteDasher** [comment]
Here, according to what is said in https://github.com/flatpak/flatpak/issues/1572, I finally understood what to do. Application icons and mime type icons are displayed correctly.

**Comment by BiteDasher** [comment]
bot, build sh.ppy.osulazer

File: x-osu.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        <comment xml:lang="zh">osu! 谱面</comment>
+        <comment xml:lang="ru">Карта osu!</comment>
+        <icon name="sh.ppy.osulazer-x-osu-beatmap"/>
+        <glob pattern="*.osz"/>
+    </mime-type>
+    <mime-type type="application/x-osu-skin">
+        <sub-class-of type="application/zip"/>
+        <comment>osu! Skin</comment>
+        <comment xml:lang="ru">Скин osu!</comment>
+        <icon name="sh.ppy.osulazer-x-osu-beatmap"/>
```
Comment (hfiguiere) [inline x-osu.xml:16]:
```suggestion
        <icon name="sh.ppy.osulazer.x-osu-beatmap"/>
```

File: x-osu.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<mime-info xmlns='http://www.freedesktop.org/standards/shared-mime-info'>
+    <mime-type type="application/x-osu-beatmap">
+        <sub-class-of type="application/zip"/>
+        <comment>osu! Beatmap</comment>
+        <comment xml:lang="hk">osu! 譜面</comment>
+        <comment xml:lang="zh">osu! 谱面</comment>
+        <comment xml:lang="ru">Карта osu!</comment>
+        <icon name="sh.ppy.osulazer-x-osu-beatmap"/>
```
Comment (hfiguiere) [inline x-osu.xml:9]:
```suggestion
        <icon name="sh.ppy.osulazer.x-osu-beatmap"/>
```

File: sh.ppy.osulazer.yaml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      - install -D -m 644 lazer-128.png /app/share/icons/hicolor/128x128/apps/sh.ppy.osulazer.png
+      - install -D -m 644 lazer-64.png /app/share/icons/hicolor/64x64/apps/sh.ppy.osulazer.png
+        # mimetypes
+      - install -D -m 644 mime-512.png /app/share/icons/hicolor/512x512/mimetypes/sh.ppy.osulazer-x-osu-beatmap.png
+      - install -D -m 644 mime-256.png /app/share/icons/hicolor/256x256/mimetypes/sh.ppy.osulazer-x-osu-beatmap.png
+      - install -D -m 644 mime-128.png /app/share/icons/hicolor/128x128/mimetypes/sh.ppy.osulazer-x-osu-beatmap.png
+      - install -D -m 644 mime-64.png /app/share/icons/hicolor/64x64/mimetypes/sh.ppy.osulazer-x-osu-beatmap.png
+      - install -D -m 644 mime-48.png /app/share/icons/hicolor/48x48/mimetypes/sh.ppy.osulazer-x-osu-beatmap.png
+      - install -D -m 644 mime-32.png /app/share/icons/hicolor/32x32/mimetypes/sh.ppy.osulazer-x-osu-beatmap.png
+      - install -D -m 644 mime-16.png /app/share/icons/hicolor/16x16/mimetypes/sh.ppy.osulazer-x-osu-beatmap.png
```
Comment (hfiguiere) [inline sh.ppy.osulazer.yaml:42]:
```suggestion
      - install -D -m 644 mime-512.png /app/share/icons/hicolor/512x512/mimetypes/sh.ppy.osulazer.x-osu-beatmap.png
      - install -D -m 644 mime-256.png /app/share/icons/hicolor/256x256/mimetypes/sh.ppy.osulazer.x-osu-beatmap.png
      - install -D -m 644 mime-128.png /app/share/icons/hicolor/128x128/mimetypes/sh.ppy.osulazer.x-osu-beatmap.png
      - install -D -m 644 mime-64.png /app/share/icons/hicolor/64x64/mimetypes/sh.ppy.osulazer.x-osu-beatmap.png
      - install -D -m 644 mime-48.png /app/share/icons/hicolor/48x48/mimetypes/sh.ppy.osulazer.x-osu-beatmap.png
      - install -D -m 644 mime-32.png /app/share/icons/hicolor/32x32/mimetypes/sh.ppy.osulazer.x-osu-beatmap.png
      - install -D -m 644 mime-16.png /app/share/icons/hicolor/16x16/mimetypes/sh.ppy.osulazer.x-osu-beatmap.png
```

File: x-osu.xml (hunk 1 above)
Comment (BiteDasher) [reply x-osu.xml]: done

**Comment by BiteDasher** [comment]
bot, build sh.ppy.osulazer

File: sh.ppy.osulazer.yaml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+          PKG_CONFIG_PATH: /app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet5/lib/pkgconfig
+    sources:
+      - nuget.json
+      - type: git
+        url: https://github.com/peppy/osu.git
+        commit: a9208838a9dcff71a73fd0ab103f1251a2817d3e
+      - type: file
+        path: osu-lazer
+      - type: file
+        path: icons/lazer-512.png
```
Comment (barthalion) [inline sh.ppy.osulazer.yaml:71]: JFYI, unless also used internally, 512x512 icon would be automatically downscaled to smaller ones. Not sure about MIME icons.

Comment (BiteDasher) [reply sh.ppy.osulazer.yaml:71]: I don't know anything about the resize, but now I looked at my system, I don't have 512x512 application icons. Do you think it's worth deleting the one on 512?

Comment (BiteDasher) [reply sh.ppy.osulazer.yaml:71]: At the same time, icons of mime types on 512x512 exist. I think maybe it's for HI-DPI?

**Comment by peppy** [comment]
uh.....

**Comment by peppy** [comment]
Here I was thinking the fact that the checklist was clearly not completed would mean this wouldn't see the light of day, but I guess here we are.

Please remove this completely. The creator of this pull request was explicitly asked to not continue with this path https://github.com/ppy/osu/discussions/15278#discussioncomment-1529371.

**Comment by BiteDasher** [comment]
> Here I was thinking the fact that the checklist was clearly not completed would mean this wouldn't see the light of day (which is the reason I didn't comment on this PR - it seemed unnecessary), but I guess here we are.
> 
> Please remove this completely. The creator of this pull request was explicitly asked to not continue with this path https://github.com/ppy/osu/discussions/15278#discussioncomment-1529371. None of the last three checkboxes are valid and this should not have been accepted.
> 
> If necessary I can provide legal means to remove this using trademark law, but hopefully that won't be required.

:(

I wrote in the description so that you do not receive bug reports, what else is the problem?  I just wanted to make the game more accessible on Linux ...

**Comment by hfiguiere** [comment]
> Here I was thinking the fact that the checklist was clearly not completed would mean this wouldn't see the light of day (which is the reason I didn't comment on this PR - it seemed unnecessary), but I guess here we are.

One would think that given the license, this wouldn't be a problen.

> If necessary I can provide legal means to remove this using trademark law, but hopefully that won't be required.

This is not really the open source spirit. Ultimately it's your own choice but you are sending a very mixed message. And just to clarify, Flathub isn't a commercial "for profit" store (it's not Steam, GOG, or Epic). It is a community managed software repository run with the help of a few sponsors.

It is sad that there can't be that kind of cooperation.

Also to answer to "use the provided packages" on the linked discussion, what package from upstream shall I use to run this on my RaspberryPi 400?

(as a note:  I can't remove it, but an admin like @barthalion can do it)

**Comment by peppy** [comment]
A license does not give permission to impersonate a domain identifier, brand, or trademark. If a third party wishes to publish the game such as what is done here, there is no issue with it, but it would be wise to first rename the product and release it as an unofficial or rebranded fork. I presume the intention of the checklist in the PR template on this repository is to ensure due diligence is taken in ensuring this happens.

I am sorry if it seems like we are sending mixed messages by not wanting our application to be listed without our permission on websites/stores, but everyone on the team agrees that this feels very wrong and against our intentions. The copy, icons and associated screenshots are not representative of how we would publish this app were we to do so.

We are not ready to do this ourselves, outside the means which we are currently distributing it. This may change in the future as we become more comfortable with the release state of the application, but that's not where we are.

> Also to answer to "use the provided packages" on the linked discussion, what package from upstream shall I use to run this on my RaspberryPi 400?

You should be able to build from source without issue. But keep in mind that some online functionality will be unavailable if you do this.

Thanks for understanding.

**Comment by barthalion** [comment]
Hey Dean,

This is always a problematic situation whether something is distributable or not, including trademarks. IANAL (and I would rather not involve legal staff), but in my opinion the use case here falls into nominative fair use, even with a registered trademark. That being said, that's not a very productive angle.

Unless you are very strict about it, maybe we can wait to see how it goes, and remove it only if the Flathub version starts causing unnecessary bug reports? Alternatively, would it be still problematic to distribute your software on Flathub if it re-used the upstream-provided AppImage file? Would it change anything if said file would be downloaded during installation instead of redistributed directly? There's also Flathub Beta, which is a separate repository not displayed on Flathub website.

if none of these options are acceptable (but I'd really love to discuss if we can do anything to change your stance first), you can effectively remove it from Flathub by adding a `flathub.json` file with `end-of-life` key like described at: https://github.com/flathub/flathub/wiki/App-Maintenance#end-of-life. I can handle it tomorrow morning as well.

**Comment by peppy** [comment]
Honestly I don't have the bandwidth to even look into what alternative options exist at the moment. I spend every waking hour working on the game, and am just trying to avoid third parties advertising / using / distributing it in ways that we didn't intend (or don't yet intend).

It looks to also be being distributed with online functionality enabled, but likely in a broken state (non-official releases won't properly connect to multiplayer / leaderboards and throw errors instead).

If it is to remain, I would suggest at very least that you change the domain name `sh.ppy.osu` and package name to reflect the fact it is unofficial. My hope would be that it is removed completely until a point at which we were looking to list it ourselves, but it seems that isn't how things work here so will revisit at a further date with an EOL or otherwise.

**Comment by BiteDasher** [comment]
(@barthalion when the discussion is over, please reply what I have to do)

**Comment by barthalion** [comment]
To summarize, the alternative options I mention are what we usually use for commercial software. In the first, it means official AppImage would be extracted and redistributed through our CDN without changes.
The second one means downloading the AppImage as-is, so that we don't redistribute anything but description/screenshots and validation metadata to ensure checksum and file size matches. This is what is being used for e.g. Slack, Spotify, Edge, JetBrains software and is considered legally "clean" and accepted by these vendors.

We can discuss it when you have spare cycles and remove/unlist OSU for the time being.

**Comment by peppy** [comment]
If the package is simply redistributing our AppImage I would take less issue.

I would hope the metadata is also taken from one of the available official sources rather than weirdly rewritten. Also the app is named wrong (it is not "osu!lazer" but "osu!").

The message on Jetbrains product listings about not being affiliated also looks to read better than the one used here, so that would be an improvement.

**Comment by hfiguiere** [comment]
> If the package is simply redistributing our AppImage I would take less issue.

Is there a aarch64 appimage?

**Comment by gagahpangeran** [comment]
> > If the package is simply redistributing our AppImage I would take less issue.
> 
> Is there a aarch64 appimage?

I'm only normal contributor in osu!(lazer) project, but here's some info.

There's no aarch64 (arm64) build for appimage, because osu-framework, the game framework that this app use, doesn't even support aarch64. Some native libs in osu-framework itself need aarch64 support (e.g. libass).

**Comment by marcan** [comment]
Reminder that open source software is about copyright licensing; trademark usage rights are a different issue, and it is common to have open source software that only allows its official branding to be used with official builds (a prime example would be Firefox).

I find it quite appalling that @BiteDasher was asked not to do this and decided to do so anyway, against the wishes of @peppy and the rest of the team. Licensing issues aside, that is just common courtesy in open source communities.

**Comment by BiteDasher** [comment]
> Reminder that open source software is about copyright licensing; trademark usage rights are a different issue, and it is common to have open source software that only allows its official branding to be used with official builds (a prime example would be Firefox).
> 
> I find it quite appalling that @BiteDasher was asked not to do this and decided to do so anyway, against the wishes of @peppy and the rest of the team. Licensing issues aside, that is just common courtesy in open source communities.

Okay, already 3 people have attacked me.  I did everything necessary to prevent the developer from receiving unwanted bug reports, since this is the main reason here, but there are also others reasons by the peppy messages above.  I'll leave it to the flathub admins whether to remove this package or not. Just keep in mind, I wasn't trying to screw up the developer.  Although I understand that he said that it was undesirable to do this, I thought that the inscription about the unofficial build and issues link should be enough. But, since more and more people come and come, I think I will soon add a message about the `end-of-life` of the package.

**Comment by barthalion** [comment]
> Reminder that open source software is about copyright licensing; trademark usage rights are a different issue, and it is common to have open source software that only allows its official branding to be used with official builds (a prime example would be Firefox).

It's not exactly a good example. Mozilla never had a problem with Firefox being distributed under that name as long as source code is not modified. Trademark usage is really more nuanced than this.

> that is just common courtesy in open source communities

I don't want to derail this discussion any further, but respect goes both ways. The submitter has spent their free time properly packaging your software in the hope of making it easier to discover and install. It's not as if it has been done out of disrespect, mind you.

<hr>

For now, I have marked OSU as end-of-life manually to avoid any further confusion. I will let @BiteDasher decide if he's still interested in reworking OSU packaging to use AppImage directly or indirectly. The GitHub repository is left intact, and any desired changes can still be pushed there.

**Comment by peppy** [comment]
Thanks @barthalion, that seems like the best short term solution. Once we get to the point of releasing the cross-platform version of our game to our mainstream audience we may revisit package submission officially. Alternative, if @BiteDasher or anyone else is look to do this in a way we won't be adverse to, I'll summarise what I've mostly already said above as criteria:

- The package should be named "osu!" not "osu!lazer". lazer is a codename and is not intended to be shown most places visible to a user. It will also change as we get further into development
- All copy (text) and image resources should be sourced from an official means (ie. https://osu.ppy.sh or https://osu.ppy.sh/wiki/en/Main_Page at very least). Two examples in the current package are the main game summary, which doesn't sit right with me, and the main banner, which contains artwork from one of our fanart contests which has user submissions that are not licensed for use like this.
- The AppImage should be used, preferably without modifications (this package changes the visible release stream in game to "flathub", which feels awkward and will likely break as we further build on that area of the UI).

**Comment by hfiguiere** [comment]
@barthalion and if you can please yank the aarch64 build so that despite the deprecation notice nobody ever installs it as it is broken. See https://github.com/flathub/sh.ppy.osulazer/issues/3

**Comment by barthalion** [comment]
@BiteDasher I think it would be ideal to create a brand new submission with `sh.ppy.osu` ID (not `osulazer`), if you want to follow up on changes requested by Dean. Feel free to poke me for review/help, and we will also consult the new submission with OSU team, of course. I will also archive the existing repo.

@hfiguiere I deleted the aarch64 ref.

**Pull request closed without merging**